### ISSUE TITLE
Feature: add primop to collect & present warnings at the end of building

### ISSUE THREAD

**Issue by moretea** [NixOS/nix#1197]
As an alternative to `abort` in https://github.com/NixOS/nixpkgs/pull/22096, it would be perfectly acceptable to have nix-build print a list of warnings at the end of the evaluation.

I image that a `log` primop could be used to build this functionality.

```nix
builtins.log { showToUser = true; severity = "warning"; message = "Your nixpkgs is deprecated";}
```

```
$ nix-build ...
<some build output>
Warning: your nixpkgs is deprecated
<more build output>

There were some warnings during building:
- Warning: your nixpkgs is deprecated
/nix/store/asdf-pkg-1.0/
```

```nix
builtins.log { showToUser = true; severity = "critical"; message = "Your nixpkgs is unsupported";}
```

```
$ nix-build ...
<some build output>
CRITICAL: your nixpkgs is unsupported
<more build output>

There were some CRITICAL errors during building:
- CRITICAL: your nixpkgs is unsupported
/nix/store/asdf-pkg-1.0/
```

**Comment by copumpkin** [comment]
See also #749

**Comment by copumpkin** [comment]
We'd need `builtins.log` to take another argument for it to work in a purely functional setting, but otherwise something like this would be very nice.

**Comment by moretea** [comment]
Oh yes, you're completely right. I had the same semantics as `builtins.trace` in mind ;) Will update.

**Comment by copumpkin** [comment]
I do like your approach more than #1000 for a couple of reasons:

1. Because it only produces the messages at the end (presumably externally to the language), it doesn't break purity
2. We can produce multiple messages at once so you don't get the usual pain from warnings and errors "trickling out" one at a time, which is always annoying in programs

cc @shlevy @edolstra @domenkozar

**Comment by edolstra** [comment]
What would `showToUser = false` do?

**Comment by edolstra** [comment]
@copumpkin The problem with delaying messages to the end is that you don't see them at all in case of a crash, infinite loop, etc.

**Comment by moretea** [comment]
I guess that `showToUser = false`  should be the default; it would only print the log entry inline, and not duplicate it at the end. You'd only use `showToUser` in cases where it's important that the end user sees this message.

This duplication might be annoying. As @copumpkin suggested, we could only print them only at the end of the `nix-build` output. We could add a flag   `--show-logs` to force the logs to be printed directly when created, just like `--show-trace` works today.

**Comment by copumpkin** [comment]
@edolstra yeah, printing at the end doesn't seem fundamental, but might be useful. The main thing would be to provide the mechanism.

**Comment by moretea** [comment]
I'm willing to implement this once we reach consensus (read: @edolstra is OK with all this)

**Comment by tomberek** [comment]
Closing as stale. Re-open if needed.
